### PR TITLE
App Check: store App Attest artifact per key ID

### PR DESCRIPTION
**Pull request by maksymmalyhin**
- make sure an artifact generated for an old App Attest key is not used to generate assertions

#no-changelog

### CODE REVIEW

**Review by paulb777 — APPROVED** [review]

File: FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h [hunk 1]
```diff
@@ -24,16 +24,20 @@ NS_ASSUME_NONNULL_BEGIN
...
 
-/// Set the artifact.
+/// Set the artifact. An artifact previously set for *any* key ID will replace the new one with the
+/// new key ID. The storage always stores as single artifact.
 /// @param artifact The artifact data to store. Pass `nil` to remove the stored artifact.
+/// @param keyID The App Attest key ID used to generate the artifact.
 /// @return An artifact that is resolved with the artifact data passed into the method in case of
 /// success or is rejected with an error.
-- (FBLPromise<NSData *> *)setArtifact:(nullable NSData *)artifact;
+- (FBLPromise<NSData *> *)setArtifact:(nullable NSData *)artifact forKey:(NSString *)keyID;
```
Comment (ncooke3) [inline FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h:33]: Should the keyID have the `nonnull` attribute that is specified in the method's implementation? Does adding it here give any more type/nullability safety? ... I was wondering about this and unsure

File: FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h [hunk 2]
```diff
@@ -24,16 +24,20 @@ NS_ASSUME_NONNULL_BEGIN
 /// Check token obtained with App Attest provider.
 @protocol FIRAppAttestArtifactStorageProtocol <NSObject>
 
-/// Set the artifact.
+/// Set the artifact. An artifact previously set for *any* key ID will replace the new one with the
+/// new key ID. The storage always stores as single artifact.
```
Comment (ncooke3) [inline FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h:28]: nit(typo): I think it should be- The storage always stores **a** single artifact.

File: FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h [hunk 3]
```diff
@@ -24,16 +24,20 @@ NS_ASSUME_NONNULL_BEGIN
 /// Check token obtained with App Attest provider.
 @protocol FIRAppAttestArtifactStorageProtocol <NSObject>
 
-/// Set the artifact.
+/// Set the artifact. An artifact previously set for *any* key ID will replace the new one with the
```
Comment (ncooke3) [inline FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h:27]: Should it say "... will **be replaced by** the new one..."

File: FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestStoredArtifact.m [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  [coder encodeObject:self.artifact forKey:kArtifactKey];
+  [coder encodeInteger:self.storageVersion forKey:kStorageVersionKey];
+}
+
+- (nullable instancetype)initWithCoder:(nonnull NSCoder *)coder {
+  NSInteger storageVersion = [coder decodeIntegerForKey:kStorageVersionKey];
+
+  if (storageVersion < kStorageVersion) {
+    // Handle migration here when new versions are added
+  }
```
Comment (ncooke3) [inline FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestStoredArtifact.m:57]: i like this, smart!

**Review by ncooke3 — APPROVED** [review]
LGTM

File: FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h (hunk 1 above)
Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h:33]: All variables placed between `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END` are marked as `nonnull`, so we need to specify nullability explicitly only for nullable variables/parameters. Without `NS_ASSUME_NONNULL_...` we would have to add the nullability modifiers for each variable.

**Comment by maksymmalyhin** [comment]
I'm going to merge it as the last commit looks safe and CI passed for the previous one.

File: FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h (hunk 1 above)
Comment (ncooke3) [reply FirebaseAppCheck/Sources/AppAttestProvider/Storage/FIRAppAttestArtifactStorage.h:33]: Ohhh I see. I understand that (and the purpose of `NS_ASSUME_NONNULL_BEGIN`) much better now– thank you!

**Pull request merged**